topdown/net.go

```diff
@@ -0,0 +1,51 @@
+package topdown
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+type lookupIPAddrCacheKey string
+
+func builtinLookupIPAddr(bctx BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	name, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	key := lookupIPAddrCacheKey(name)
+	if val, ok := bctx.Cache.Get(key); ok {
+		return iter(val.(*ast.Term))
+	}
+
+	addrs, err := net.DefaultResolver.LookupIPAddr(bctx.Context, string(name))
+	if err != nil {
+		var derr *net.DNSError
+		if errors.As(err, &derr) && derr.Temporary() {
+			return Halt{
+				Err: &Error{
```

Comment:
We probably need to set the AST location on the inner error.

Reply:
```
open-policy-agent/opa-main % ./opa_darwin_amd64 eval -fpretty --strict-builtin-errors 'net.lookup_ip_addr("exmaplesishf.org")'
1 error occurred: net.lookup_ip_addr("exmaplesishf.org"): eval_builtin_error: net.lookup_ip_addr: lookup exmaplesishf.org: no such host
open-policy-agent/opa-main % ./opa_darwin_amd64 eval -fpretty --strict-builtin-errors 'http.send({"url":"https://exmaplesishf.org", "method":"get"})'
1 error occurred: http.send({"url":"https://exmaplesishf.org", "method":"get"}): eval_builtin_error: http.send: Get "https://exmaplesishf.org": dial tcp: lookup exmaplesishf.org: no such host
open-policy-agent/opa-main % ./opa_darwin_amd64 eval -fpretty 'http.send({"url":"https://exmaplesishf.org", "method":"get"})'                       
undefined
open-policy-agent/opa-main % ./opa_darwin_amd64 eval -fpretty  'net.lookup_ip_addr("exmaplesishf.org")'                       
undefined
```
removed the halt error, it now behaves more like http.send ☝️

Reply:
Re-introduced the halt error for context-related errors.